### DISCUSSION TITLE
How to apply "mode" function to coarsen object

### DISCUSSION THREAD

**Discussion by chfite** [pydata/xarray#4960] [Q&A]
I was wondering if it is possible to apply "mode" to a coarsened array object, instead of mean, min, or max, etc. I know that numpy does not have a mode function, and so I was wanting to use the scipy.stats.mode function and apply it. I've looked at xr.apply_ufunc() but have had little luck getting it to work taking the mode of the coarsened object.

For example, lets say I have a dataarray with dims (lat, lon) and shape (415, 725) and I want to coarsen it every 5 grid cells. I was going to do datarray.coarsen(lat=5, lon=5) but am not sure the best way to get the mode of my values that are "grouped" in each larger gridcell, which is what I need.

**Comment by dcherian** [comment]
`coarsen` very recently gained a `.reduce` method which will be included in an upcoming release very soon. You could test it out on the master branch in the mean time.

**Reply by spencerkclark** [reply]
Indeed, this is a good application for the forthcoming `.reduce` method.  That said, even with `.reduce` this is somewhat of a tricky (but solvable) problem, because `scipy.stats.mode` does not have the same interface as traditional NumPy reduction functions like `mean` or `sum`.  The main challenge is wrapping it in such a way that you can do a mode reduction along multiple axes at once by passing a tuple of axes as an argument ([currently](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mode.html) it can only take a single axis).  [NumPy's private `_ureduce` function](https://github.com/numpy/numpy/blob/a14c41264855e44ebd6187d7541b5b8d59bb32cb/numpy/lib/function_base.py#L3464-L3516) is a good illustration for how to do this; I would be curious to know if there is a public equivalent of that in NumPy or some other library.
